lightning/src/offers/invoice.rs

```diff
@@ -340,18 +340,29 @@ macro_rules! invoice_builder_methods { (
 	$self: ident, $self_type: ty, $return_type: ty, $return_value: expr, $type_param: ty $(, $self_mut: tt)?
 ) => {
 	pub(crate) fn amount_msats(
-		invoice_request: &InvoiceRequest
+		invoice_request: &InvoiceRequest, custom_amount_msats: Option<u64>
 	) -> Result<u64, Bolt12SemanticError> {
-		match invoice_request.amount_msats() {
-			Some(amount_msats) => Ok(amount_msats),
-			None => match invoice_request.contents.inner.offer.amount() {
-				Some(Amount::Bitcoin { amount_msats }) => {
-					amount_msats.checked_mul(invoice_request.quantity().unwrap_or(1))
-						.ok_or(Bolt12SemanticError::InvalidAmount)
-				},
-				Some(Amount::Currency { .. }) => Err(Bolt12SemanticError::UnsupportedCurrency),
-				None => Err(Bolt12SemanticError::MissingAmount),
-			},
+		if invoice_request.amount_msats().is_some() && custom_amount_msats.is_some() {
+			return Err(Bolt12SemanticError::UnexpectedAmount);
+		}
+
+		if let Some(amount_msats) = invoice_request.amount_msats() {
+			return Ok(amount_msats);
+		}
+
+		if let Some(custom_amount_msats) = custom_amount_msats {
+			return Ok(custom_amount_msats);
+		}
```

Comment:
It should probably be an error if we reach here and the offer has an amount denominated in bitcoin. That is the custom amount should only be applicable for currency-denominated offers. Not offers either denominated in bitcoin or with no amount. Though I _think_ in the latter case we'll fail to parse the invoice request. May want to double check that.

Reply:
Makes sense!
Updated the check to make sure custom amount is provided only when an `Offer` is in Currency denomination and `InvoiceRequest`'s amount is absent in [pr3318.03](https://github.com/shaavan/rust-lightning/commits/pr3318.03)!

Also, I [confirmed](https://github.com/lightningdevkit/rust-lightning/blob/66fb5200986e0c145a16d74b04e471d9cdd08eb5/lightning/src/offers/invoice_request.rs#L1170-L1172) that we would fail to parse the `InvoiceRequest` if both offer's amount and IR's amount is absent.

Thanks a lot for the pointer!